Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+//          if (rand() / RAND_MAX > prob)
+//            r = target; // Force a skip
+//        } while (r == target); // Don't return the target.
+//        return r;
+//      }
+      /**
+        * Randomly choose a word for negative sampling.  This does not
+        * respect the probability distribution of the words in the corpus,
+        * but is adequate and much more efficient.
+        */ 
+      uint32_t getNegative(uint32_t target)
+      {
+        uint32_t r;
+        do
+        {
+          r = rand() % wc_;
+        } while (r == target);
+        return r;
+      }
+      /**
+        * Calculate the neural network Sigmoid function for forward
+        * propagation.
+        */
+      double sigmoid(double x)
+      {
+        //return x;
+        double y;
+        if (x > 25)
+          y = 1.0;
+        else if (x < -25)
+          y = 0.0;
+        else
+          y = 1.0 / (1.0 + exp(-x));
+        return y;
+      }
+      /** Calculate the loss of a given sample.
+        * Uses linear loss rather than log loss as it
+        * is easier to evaluate and understand.
+        * label is either 1 or 0, and represents to
+        * target value.
+        */
+      float calcLoss(float score, uint32_t label)
+      {
+        float loss;
+        if (label == 1)
+          loss = 1 - score;
+          //loss = -log(score + .00001);
+        else
+          loss = score;
+          //loss = -log((1-score) + .00001);
+        return loss;
+      }
+      /**
+        * Forward and Back Propagation to adjust weights.
+        * Processes a single training sample.
+        * For each sample, processes the positive case
+        * (i.e. the sample) as well as a set of randomly
+        * selected negative samples.  Saves the weights
+        * and the weight updates (deltas).
+        * First, propagate the inputs (the context variables)
+        * to the output to compute the output (score).
+        * Then back-propagate errors (i.e. score - label)
+        * back to the output weights and the input weights.
+        */
+      void fbPropagate(tp& train)
+      {
+        //assert(train.main > 0 & train.main <= wc_);
+        //if (train.main < 1 || train.main > wc_)
+        //  return;
+        uint32_t main = train.main - 1; // Map 1-based ECL index to 0-based
+        uint32_t contxCount = train.contxCount;
+        uint32_t * contx = train.contx;
+        double contxCountInv = 1.0 / (double)contxCount;
+        // FORWARD PATH ***
+        uint32_t j, i, n, windx, target, label;
+        double cum, score, alpha, grad, wUpdate;
+        // Zero the hidden layer.
+        for (uint32_t i = 0; i < dim_; i++)
+          hidden_[i] = 0;
+        // Update the hidden layer outputs based the context words.
+        for (j = 0; j < contxCount; j++)
+        {
+          uint32_t cx = contx[j] - 1; // Map 1-based ECL index to 0-based
+          for ( i = 0; i < dim_; i++)
+          {
+              windx = toFlatIndx(0, cx, i);
```

Comment:
if you are optimizing the code it would be good to check exactly what code is generated for this.  but not something for this PR

Reply:
deferred